lredis/commands.lua

```diff
@@ -1,79 +1,150 @@
 end
 
-function methods:ping()
-	local resp = self:pcall("PING")
-	return handle_ok_or_err(resp)
+function base_methods:ping()
+	return self:call_ok_or_err(nil, "PING")
 end
 
-function methods:client_pause(delay)
+function base_methods:client_pause(delay)
 	local milliseconds = string.format("%d", math.ceil(delay*1000))
-	local resp = self:pcall("client", "pause", milliseconds)
-	return handle_ok_or_err(resp)
+	return self:call_ok_or_err(nil, "client", "pause", milliseconds)
 end
 
-function methods:subscribe(...)
+function base_methods:subscribe(...)
 	self:start_subscription_mode("SUBSCRIBE", ...)
 end
 
-function methods:unsubscribe(...)
+function base_methods:unsubscribe(...)
 	self:start_subscription_mode("UNSUBSCRIBE", ...)
 end
 
-function methods:punsubscribe(...)
+function base_methods:punsubscribe(...)
 	self:start_subscription_mode("PUNSUBSCRIBE", ...)
 end
 
-function methods:psubscribe(...)
+function base_methods:psubscribe(...)
 	self:start_subscription_mode("PSUBSCRIBE", ...)
 end
 
-function methods:multi()
-	local resp = self:call("MULTI")
-	local ret = handle_ok_or_err(resp, 2)
-	self:start_transaction()
-	return ret
+--[[
+--	Client methods
+--]]
+
+-- execute the MULTI command and return a new transaction object if successful.
+function client_methods:multi()
+	if self.transaction_lock and not self:in_coroutine() then
+		error("Transaction in progress and cannot wait -- not in a coroutine", 2)
+	end
+
+	while self.transaction_lock do
+		self.transaction_lock:wait()
+	end
+	self:create_transaction_lock()
+
+	local transaction = new_transaction(self)
+	transaction:call_ok_or_err(nil, "MULTI")
+
+	return transaction
 end
 
-function methods:exec()
-	local resp = self:call("EXEC")
-	self:end_transaction()
-	return resp
+function new_client(client)
+	return setmetatable(client, client_mt)
 end
 
-function methods:discard()
-	local resp = self:call("DISCARD")
-	local ret = handle_ok_or_err(resp, 2)
-	self:end_transaction()
-	return ret
+--[[
+--	Transaction methods
+--]]
+
+function transaction_methods:end_transaction()
+	if self.in_transaction then
+		self.client.subscribes_pending = self.subscribes_pending
+		self.client.subscribed_to = self.subscribed_to
+
+		self.in_transaction = nil
+		self.client:destroy_transaction_lock()
+	end
+end
+
+function transaction_methods:call(func, ...)
+	if not self.in_transaction then
+		error("Transaction no longer valid", 2)
+	end
+
+	local resp = self:call_no_err(func, ...)
+	func = func:upper()
+	if func == "EXEC" or func == "DISCARD" or resp.err then
+		self:end_transaction()
+	end
+
+	if resp.err then
+		error(resp.err, 2)
+	end
+	return (resp.ok and resp) or resp.data
 end
 
-return methods
+function transaction_methods:exec()
+	return self:call("EXEC")
+end
+
+function transaction_methods:discard()
+	return self:call_ok_or_err(nil, "DISCARD")
+end
+
+function new_transaction(client)
+	local transaction = {
+		client = client,
+		socket = client.socket,
```

Comment:
don't copy these out: just refer to them via `client.socket` at use-sites.

Reply:
We can't do that without exposing the difference between a transaction and a client to the lower-level code.  That would require much larger and unnecessary changes.  The lower level has no need to know the difference, so we must pass it something that has the same interface as a client and that is what this code does: it makes a transaction look like a client to the lower level.